packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -931,6 +944,49 @@ Hardhat Network's forking functionality only works with blocks from at least spu
     await this._txPool.updatePendingAndQueued();
   }
 
+  public async setAccountBalance(
+    address: Address,
+    newBalance: BN
+  ): Promise<void> {
+    const account = await this._stateManager.getAccount(address);
+    account.balance = newBalance;
+    await this._stateManager.putAccount(address, account);
+    this._irregularStatesByBlockNum.set(
+      (await this.getLatestBlock()).header.number.toString(),
+      await this._stateManager.getStateRoot()
+    );
+  }
+
+  public async setAccountCode(
+    address: Address,
+    newCode: Buffer
+  ): Promise<void> {
+    await this._stateManager.putContractCode(address, newCode);
+  }
+
+  public async setAccountNonce(address: Address, newNonce: BN): Promise<void> {
+    const account = await this._stateManager.getAccount(address);
+    if (newNonce.lt(account.nonce)) {
+      throw new InvalidInputError(
+        "New nonce must not be smaller than the existing nonce"
```

Comment:
Let's add the current nonce and the used nonce as part of the message, that's usually super helpful. Something like: "New nonce (3) must not be smaller than the existing nonce (5)"

Reply:
Addressed in https://github.com/nomiclabs/hardhat/pull/1435/commits/5a0631dac3f3f4ab0a1ab7c8c1b89aee6e2c3207